### ISSUE TITLE
System Speed

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#438]
In GitLab by @kdjstudios on Jul 28, 2017, 13:49

Gary wrote: http://www.servicedesk.answernet.com/profiles/ticket/2017-07-28-59652/conversation

The system speed is terribly slow when running reports.  I tried to run calpine's ledger and billing history and they were really slow and then we tried to do a pdf and it took forever.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 5, 2017, 09:41

This is the Aavaz spreadsheet with run times of each report. 

[Report_Definitions_from_Aavaz.xlsx](/uploads/4bca27e7d1b85c1e4c201afa891bc40a/Report_Definitions_from_Aavaz.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Dec 5, 2017, 12:56

Another benchmark is that running the Account Ledger for Calpine Energy Solutions, LLC under the NCSM site took 1:57 to display.

**Comment by kdjstudios** [comment]
In GitLab by @abakshi on Dec 13, 2017, 06:36

assigned to @smasih

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 14, 2017, 07:46

I was not able to spend more time on this today but, will continue with it tomorrow.

Account Ledger is calculated on the fly, which is perhaps the biggest reason that this report is slow.

Besides, there are at least 5 loops/traversing in the ledger code if we can reduce this maybe we can make it a bit faster.

We are using ruby to sort and filter over received chunk of data, but we can optimize this by improving the mongo query in the ledger code.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 15, 2017, 05:27

@kdjstudios 
@tim.traylor 


Didn't want to directly work on Answernet or merge for obvious reasons. So we chose dev server.

Keener Communication has account "LONG & FOSTER" and it's Ledger/statement **Report takes 24s.**

Firstly we had to identify the bottle necks:
* Is it App (ruby/rails)
* Is it DB (mongo) / query taking too long
* or it's just the code.

As per my findings we can improve by adding better queries and escaping the loop.  
Ledger report is being generated on the fly, we get all the account payments and invoices and sort them on basis of their payment/invoice dates. 

As a resolution (Suggested by Garry) we can save the report directly into db and write backend job to weekly update the ledger report in db for the account.

Alternatively we can try to improve the code and query. **We made few changes in Ledger Report and its now taking 8-9s to load instead of 24s.** 

Changes were very simple we just reduced recursive direct db calls by adding variables so instead of directly searching in db it searches in its place holder.

**Comment by kdjstudios** [comment]
In GitLab by @abakshi on Dec 18, 2017, 06:49

Check Parameters
* ulimit - limit resource usage for a process in terms of priority, threads, files, connections
* Lock Time%
* Storage Engine - DB size, Index size, fragmentation
* Open connections%
* Profiling - Re-index, db stats, collection stats, Slow queries

Tools
* [Ganglia](https://github.com/quiiver/mongodb-ganglia/blob/master/gmetric-mongodb.py) - operations per second, memory usage, btree statistics, master/slave status and current connections
* [Nagios](https://github.com/mzupan/nagios-plugin-mongodb/blob/master/check_mongodb.py) - Open Connection%, Memory Usage, index miss ratio, # of databases and # of collections, DB size, Index size, Queries/second

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 19, 2017, 07:45

* Tim asked to check if there was a background job for reindexing.

````bash
[sabilling@answernet home]$ crontab -l
MAILTO=blanca.mancilla@sabilling.com
MAILTO=arosenfeld@rubysavvy.com
01 00 1 * * /home/sabilling/jobs/run_ar_snapshot.sh
01 01 * * * /home/sabilling/jobs/run_check_aging.sh
20 4 * * * /opt/sabilling/config/unicorn_init.sh restart
````

I was not able to find any background job as a `sabilling user`. There were also no jobs listed in `/home/sabilling/jobs` related to re-indexing of mongodb.

**Is there somewhere else I was supposed to look ..??**

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 19, 2017, 08:37

Thanks for the key points @abakshi 



> Profiling - Re-index, db stats, collection stats, Slow queries

I was not able to collect any data by doing profiling on dev server, I will try it on answernet server.


In generic my suggestion will be to upgrade our mongodb as currently we are running `2.6` and migrating from `2.6` to `3.0` is not that hard. But after doing this we can achieve 2x or 3x speed in our read or write queries.

Besides we have to manually check indexing of most frequently used model, to check if we are fully utilizing mongo.

> Still working on to figure out if there are any hardware changes required

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 21, 2017, 07:25

I ran db profiling on Answernet server for one day and collected data **Before running re-indexing manually**

* Number of slow queries
 * invoices - 779
 * payments - 27
 * accounts - 61
 * customers - 2
 * credit_card_transactions - 3
* Slowest Collections
 * payments
 * invoices
 * customers
 * accounts
 * credit_card_transactions

I will post the data of another day **after running re-indexing manually**

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Dec 22, 2017, 03:46

I did not clean up the old logged data due to which new data stacked up on the old data, I have cleaned the old data to get more accurate results, so I need one more day to record the data.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jan 5, 2018, 24:14

Data after re-indexing collections.

* Number of slow queries
 * invoices - 234
 * payments - 566
 * accounts - 114
 * customers - 17
 * credit_card_transactions - 31
 * transactions/billing - 8

* Slowest Collections in descending order 
 1. payments
 1. credit_card_transactions
 1. accounts
 1. customers
 1. transactions/billing detail 
 1. invoices

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 5, 2018, 09:00

@tim.traylor 

If I am reading this correctly then reindexing only improved the Invoices, everything else got worse at a extremely high rate right?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jan 5, 2018, 23:39

@kdjstudios 

reindexing always increases the speed, maybe the day we collected the data was a busy day.. I will post more data to get the average.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 7, 2018, 24:25

@smasih 

Thank you for clarifying. really hoping that we didn't slow the system down more.  :-(

**Comment by kdjstudios** [comment]
In GitLab by @abakshi on Jan 12, 2018, 24:20

@smasih 

Here is the quick reference to keep multiple mongodb versions:
https://stackoverflow.com/questions/43411189/install-multiple-versions-of-mongodb-on-linux

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 18, 2018, 06:36

Hi [Kyle] (@kdjstudios ) & [Tim] ( @tim.traylor ),   
We have succesfully upgraded the MongoDB from 2.6 to 3.0  :clap:   
There has been increase in speed.  
On fetching `14555` accounts new Mongo added 14.67x speed.
Similarly, on fetching `345638` invoices new Mongo added 2.7x speed.  
We are still keeping it under observation & will release it on next milestone with [Tim's] (@tim.traylor ) approval.  

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 18, 2018, 16:39

This is amazing!!! Way to Go! I like the speed increase multiplicities. Almost three times as fast. That is great. Agreed on waiting for @tim.traylor approval for release.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 22, 2018, 06:00

@kdjstudios Piyush is going to note some test scenarios here that he is performing, and we will want to complement that with some testing of different things.  If all goes well, we can release in the next cycle in 2 weeks.  Thx, Tim

**Comment by kdjstudios** [comment]
In GitLab by @abakshi on Jan 22, 2018, 06:22

[Piyush](@pchaudhary) is currently focused on running all the reports. After this, he can perform overall coverage testing on following functionalities:
* Site specific - creation and modifications
* Account - creation and modifications
* Run billing cycle for key sites - Billings MT, NCSM
* Invoice generation, reversing
* Human readable
* Rebillable

@kdjstudios @tim.traylor 

Please update if you want to add any further test to ensure that we get the maximum code coverage before we can confidently release it in production.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 22, 2018, 10:22

In addition to the above the main things I foresee being used alot and is where we will want to test for speed increases is the following areas:

* Site edit page load time

* Report run time 

* Report download time - it seems like when I download a report a reruns the report to download it? Example Aging

* Account ledger - the new account ledger is certainly faster

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Feb 9, 2018, 03:32

assigned to @pchaudhary and unassigned @smasih

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Feb 9, 2018, 07:08

Hi [Kyle] ( @kdjstudios ),  
We have checked on the staging & all of the things are working perfectly. We are ready to go to upgrade the Merge server before to go live with it. 
Please approve.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 9, 2018, 08:30

@pchaudhary 

Thank you, I do not see anything else needing to be checked. 

So far it appears that all of the differences reported about Merge and Production discrepancies appear to be Human Error and not programming. I would initially approve this so far.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Feb 11, 2018, 22:33

Hi [Tim] ( @tim.traylor ),   
We need your approval to upgrade the Merge Server with Mongo 3.0.  

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Feb 27, 2018, 23:20

[wiki for detailed steps on mongo upgrade](http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/mongo-upgrade-steps)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Feb 28, 2018, 01:05

Hi [Tim] ( @tim.traylor ) and [Kyle] ( @kdjstudios ),   
We have succesfully upgraded the mongodb version from **2.6.4 to 3.0.15** in our **DEV** server.  
We have also updated a [WIKI] (http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/mongo-upgrade-steps) for the steps to be followed for the upgrade.  

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Mar 21, 2018, 24:17

Hi [Kyle] ( @kdjstudios ) & [Tim] ( @tim.traylor ),   

**Starting a discussion:** - *Please make sure to comment under a discussion area for the particular discussion.*

We are ready to upgrade the MongoDB on the Production server, just need your approval before the upgrade.
Although, if there is any concern need to have discussed on this particular one, then please do let us know.

Thx.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 30, 2018, 01:30

@kdjstudios 

No approval received from Answernet till date so moving this ticket to paused status

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 30, 2018, 08:29

@tim.traylor 

Please review and approve. I feel this is something that should only be approved by you.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 8, 2018, 15:13

@smasih @pchaudhary @tim.traylor 

Please bring this up for discussion with @tim.traylor during the morning meetings, so we can have his approval to release.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 9, 2018, 03:41

@kdjstudios 

sure, pushing it for next cycle.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 29, 2019, 13:32

@amishra @smasih @pchaudhary 

I feel the Mongo and Ruby upgrades have resolved a lot of the delay we were experiencing. Also there have been some updates to the Account Ledger too which I believe have also improved this too. 

Was there any additional updates/requests that were listed in this ticket that have not been completed?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 5, 2019, 13:56

Closing this ticket. New ticket to update Mongo, Ruby, and Rails is in #1614

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 5, 2019, 13:56

closed
